---
title: "GLOE-Seq analysis"
output: html_notebook
---

```{r message=FALSE}

library(wigglescout)
library(ggplot2)
library(reshape2)
library(dplyr)
library(ggpubr)
library(GenomicRanges)


source('../lib/lib_embed_data.v1.0.R')

data_path <- '/Volumes/DATA/DATA/GLOE/GLOE/replicates/'
atac_path <- '/Volumes/DATA/DATA/GLOE/ATAC/'

replicates <- c("WT-UT-1-1","WT-UT-1-2","WT-UT-1-3","WT-20-2-1","WT-20-2-2","WT-20-2-3","WT-24hr-3-1","WT-24hr-3-2","WT-24hr-3-3","KO-UT-4-1","KO-UT-4-2","KO-UT-4-3","KO-20-5-1","KO-20-5-2","KO-20-5-3","KO-24hr-6-1","KO-24hr-6-2","KO-24hr-6-3")
samples <- c("WT-UT-1","WT-20-2","WT-24hr-3","KO-UT-4","KO-20-5","KO-24hr-6")

build_filenames <- function(path, cond) {
  files <- paste0(path,"/GL-",cond,".hg19.bs50.bw")
  labels <- paste0("GLOE ", cond)
  return(data.frame(files=files,labels=labels))
}

build_filenames_plus <- function(path, cond) {
  files <- paste0(path,"/GL-",cond,".hg19.R1plus.bw")
  labels <- paste0("GLOE (+) ", cond)
  return(data.frame(files=files,labels=labels))
}

build_filenames_minus <- function(path, cond) {
  files <- paste0(path,"//GL-",cond,".hg19.R1minus.bw")
  labels <- paste0("GLOE (-) ", cond)
  return(data.frame(files=files,labels=labels))
}

```

## Pileup regions = 'hypersensitive breakpoints'

Given the overall low coverage/incidence of SSB in the genome, a coincidence of several SSBs in the same location (or even exact same base pair) is highly unlikely to occur by chance. Noticing that around CTCF sites, uniqe GLOE-Seq reads piled up neatly (some with exact same starting base), I developed a pipeline that looks for such pileups. There are ~40 of these regions identified, which I would call 'hypersensitive'. There may be many more but it is not possible to call them confidently.

```{r fig.width = 8, fig.height = 2}

fn <- build_filenames(data_path,samples)

p1 <- plot_bw_profile(bwfiles = fn$files,labels = fn$labels,mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F) + scale_y_continuous(limits=c(0,45))
p3 <- plot_bw_profile(bwfiles = fn$files,labels = fn$labels,mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.shuffled.bed", bin_size = 10, verbose = F) + scale_y_continuous(limits=c(0,45))

fn <- build_filenames(data_path,replicates[c(7,8,9,16,17,18)])
p2 <- plot_bw_profile(bwfiles = fn$files,labels = fn$labels,mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F) + scale_y_continuous(limits=c(0,45))

ggarrange(p1, p2, p3, ncol = 3, nrow = 1)
embed_last_plot("pileup-peaks")
```
Heatmaps of the same regions
```{r fig.width = 10, fig.height = 2}

fn <- build_filenames(data_path,samples)
p1 <- plot_bw_heatmap(bwfile = fn$files[1], mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F,zmax = 50)
p2 <- plot_bw_heatmap(bwfile = fn$files[2], mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F,zmax = 50)
p3 <- plot_bw_heatmap(bwfile = fn$files[3], mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F,zmax = 50)
p4 <- plot_bw_heatmap(bwfile = fn$files[4], mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F,zmax = 50)
p5 <- plot_bw_heatmap(bwfile = fn$files[5], mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F,zmax = 50)
p6 <- plot_bw_heatmap(bwfile = fn$files[6], mode = "center",upstream = 1000, downstream = 1000, loci = "../data/pileup/GL-WT-24hr-3.pile_up3_s10_u.uniq.bed", bin_size = 10, verbose = F,zmax = 50)


ggarrange(p1, p2, p3, p4, p5, p6, ncol = 6, nrow = 1)
embed_last_plot("pileup-heatmap")
```
```{r fig.width = 10, fig.height = 2}
histones <- c('/Volumes/DATA/DATA/GLOE/GLOE/GEO/Torres2016_H10_1.hg19.bw','/Volumes/DATA/DATA/GLOE/GLOE/GEO/Torres2016_H14_2.hg19.bw','/Volumes/DATA/DATA/GLOE/GLOE/GEO/U2OS_AcH3_ChIPseq.hg19.bw')

plot_bw_profile(bwfiles = histones[1],labels = "H1.0", mode = "start" ,upstream = 1000, downstream = 1000, loci = "../data/motif_pileup3/GL-WT-24hr-3.pileup.reads.bed", bin_size = 10, verbose = F)
```


```{r fig.width = 6, fig.height = 3}

atac <- data.frame(
    files=c("WT-NT-01.hg19.bw","WT-NT-02.hg19.bw","WT-NT-03.hg19.bw","WT-IR-01.hg19.bw","WT-IR-03.hg19.bw","CAD-NT-01.hg19.bw","CAD-NT-02.hg19.bw","CAD-NT-03.hg19.bw","CAD-IR-01.hg19.bw","CAD-IR-02.hg19.bw"),
    labels=c("ATAC WT-UT-1","ATAC WT-UT-2","ATAC WT-UT-3","ATAC WT-24h-1","ATAC WT-24h-2","ATAC KO-NT-1","ATAC KO-NT-2","ATAC KO-NT-3","ATAC KO-24h-1","ATAC KO-24h-2"))

atac$files <- paste0(atac_path,atac$files)

p1 <- plot_bw_profile(bwfiles = atac$files[c(1,4,7,9)],labels = atac$labels[c(1,4,7,9)],mode = "center",upstream = 1000, downstream = 1000, loci = "../genome/HCT116_ATACseq.hg19.bed", bin_size = 50, verbose = F) 

p2 <- plot_bw_profile(bwfiles = atac$files[c(1,4,7,9)],labels = atac$labels[c(1,4,7,9)],mode = "center",upstream = 1000, downstream = 1000, loci = "../data/motif_pileup3/GL-WT-24hr-3.pileup.reads.bed", bin_size = 50, verbose = F)

ggarrange(p1, p2, ncol = 2, nrow = 1)
embed_last_plot("pileup-atac-profile")
```

```{r fig.width = 8, fig.height = 2}

p1 <- plot_bw_heatmap(bwfile = atac$files[1],mode = "center",upstream = 1000, downstream = 1000, loci = "../data/motif_pileup3/GL-WT-24hr-3.pileup.reads.bed", bin_size = 50, verbose = F, zmax = 45) 

p2 <- plot_bw_heatmap(bwfile = atac$files[4],mode = "center",upstream = 1000, downstream = 1000, loci = "../data/motif_pileup3/GL-WT-24hr-3.pileup.reads.bed", bin_size = 50, verbose = F, zmax = 45) 

p3 <- plot_bw_heatmap(bwfile = atac$files[7],mode = "center",upstream = 1000, downstream = 1000, loci = "../data/motif_pileup3/GL-WT-24hr-3.pileup.reads.bed", bin_size = 50, verbose = F, zmax = 45) 

p4 <- plot_bw_heatmap(bwfile = atac$files[9],mode = "center",upstream = 1000, downstream = 1000, loci = "../data/motif_pileup3/GL-WT-24hr-3.pileup.reads.bed", bin_size = 50, verbose = F, zmax = 45) 

ggarrange(p1, p2, p3, p4, ncol = 4, nrow = 1)
embed_last_plot("pileup-atac-heatmap")
```
